### PR TITLE
[Adaptive] Add Binary Data Parsing For Http Request

### PR DESCRIPTION
**Pull request by zidaneymar**
fixes: #3793 

Add new response type binary data and parse the result from concrete http response content.

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/HttpRequest.cs [hunk 1]
```diff
@@ -332,6 +339,13 @@ public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc,
                     requestResult.Content = content;
                     break;
 
+                case ResponseTypes.BinaryData:
+                    // Try to resolve binary data
+                    var bytes = await response.Content.ReadAsByteArrayAsync();
+                    var encodedString = Convert.ToBase64String(bytes);
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/HttpRequest.cs:345]: >ToBase64String [](start = 48, length = 14)

Why not return as byte[] ?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/HttpRequest.cs [hunk 2]
```diff
@@ -64,7 +66,12 @@ public enum ResponseTypes
             /// <summary>
             /// Json Array of activity objects to send to the user
             /// </summary>
-            Activities
+            Activities,
+
+            /// <summary>
+            /// Binary data parsing from http response content
+            /// </summary>
+            BinaryData
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/HttpRequest.cs:74]: >BinaryData [](start = 12, length = 10)

I would call this just "Binary"

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/HttpRequest.cs [hunk 3]
```diff
@@ -332,6 +339,13 @@ public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc,
                     requestResult.Content = content;
                     break;
 
+                case ResponseTypes.BinaryData:
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/HttpRequest.cs:342]: >BinaryData [](start = 35, length = 10)

You need to add a unit test for this.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/HttpRequest.cs (hunk 1 above)
Comment (zidaneymar) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/HttpRequest.cs]: > > ToBase64String [](start = 48, length = 14)
> 
> Why not return as byte[] ?

return byte[] now, user could use LG function Base64 to parse the data.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/HttpRequest.cs (hunk 3 above)
Comment (zidaneymar) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/HttpRequest.cs]: > > BinaryData [](start = 35, length = 10)
> 
> You need to add a unit test for this.

Hi Tom, do I need to add the UT under Microsoft.Bot.Builder.Dialogs.Adaptive.Tests? I think I need a mock http client to handle all fake data... Am I right

**Comment by cleemullins** [comment]
Build isn't passing. Pushing this change out to R10.

**Comment by vishwacsena** [comment]
@zidaneymar can we add unit test and move this along to resolution?

**Comment by cleemullins** [comment]
@zidaneymar, Please resolve the merge conflicts.

File: tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/ActionTests.cs [hunk 4]
```diff
@@ -463,6 +527,12 @@ public async Task Action_HttpRequest()
                     .AssertReply("string")
                     .AssertReply("object")
                     .AssertReply("array")
+                    .AssertReply("VGVzdEltYWdl")
```
Comment (zidaneymar) [inline tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/ActionTests.cs:530]: The bytes array would be auto-base64-encoded by the sdk itself. So the result would be "VGVzdEltYWdl"

**Comment by zidaneymar** [comment]
Hi @cleemullins , I have updated the related unit tests abount 'responseTypes' in tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/ActionTests.cs, do I need to add additional unit tests

**Pull request merged**